src/corefx/System.Globalization.Native/pal_collation.c

```diff
@@ -392,31 +392,36 @@ void GlobalizationNative_CloseSortHandle(SortHandle* pSortHandle)
 
 const UCollator* GetCollatorFromSortHandle(SortHandle* pSortHandle, int32_t options, UErrorCode* pErr)
 {
-    UCollator* pCollator;
     if (options == 0)
     {
-        pCollator = pSortHandle->collatorsPerOption[0];
+        return pSortHandle->collatorsPerOption[0];
     }
     else
     {
+        options &= CompareOptionsMask;
+        UCollator* pCollator = pSortHandle->collatorsPerOption[options];
+        if (pCollator != NULL) // fast path without lock
```

Comment:
There doesn't appear to be any barriers employed here.  What guarantees that all of the referenced state is consistent?

Reply:
The full life-cycle is following:

When new `CompareInfo` is created, it's ctor calls the InitSort method:

https://github.com/dotnet/coreclr/blob/25878db0c11bfb8d52088c75ccfcc13e4c8383f2/src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.cs#L63

The `InitSort` method calls `Interop.Globalization.GetSortHandle`

https://github.com/dotnet/coreclr/blob/7e4e50ca79d9f543252af36f42b977480156a53e/src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs#L33

Which calls `ucol_open` that creates first `UCollator` and writes it to `collatorsPerOption[0]`

https://github.com/dotnet/coreclr/blob/f983ffe9aa143a15771a2a377ef27e5f2ce18b74/src/corefx/System.Globalization.Native/pal_collation.c#L365

Then anytime we need `UCollator ` for given `options` there is a `NULL` check to see if it is already initialized. If it is, we don't acquire the expensive lock (this is what my change is about).

If it's not, the actual initialization happens inside the lock and is splitted for initialzation of local variable and actual write to the array (answer to your question):

https://github.com/dotnet/coreclr/blob/f983ffe9aa143a15771a2a377ef27e5f2ce18b74/src/corefx/System.Globalization.Native/pal_collation.c#L417-L418

Later, when the sort handle is being closed, all non NULL Collators are closed and freed:

https://github.com/dotnet/coreclr/blob/f983ffe9aa143a15771a2a377ef27e5f2ce18b74/src/corefx/System.Globalization.Native/pal_collation.c#L381-L385

Does this answer your question?

Reply:
> Does this answer your question?

Thank you, but not really.  I'm asking a memory model question. You're writing lock-free code here.  One thread while holding the lock is initializing some memory and writing a pointer to that memory into an array.  Another thread without holding the lock does a hot read of that array slot, and seeing that it's not null, proceeds to use what the memory references.  I do not see any memory barriers in the code that guarantee that reads and writes aren't reordered.  What prevents the writes that initializes the collator from being reordered to after the pointer to that memory is written into the array?  What similarly prevents read reordering on the consumption side?

Reply:
can we just remove the whole lock here and use something like interlock compare exchange when writing to the array? there should be a native linux method can provide interlock compare exchange (I can search it if needed)

---
In reply to: [290819870](https://github.com/dotnet/coreclr/pull/24973#discussion_r290819870) [](ancestors = 290819870)

Reply:
Thanks for the full explanation, now I get it. I thought that you meant encapsulation.

You are right, we need memory barrier(s) here. I hardly ever use C so I am going to read some more about it and update the PR later.

Reply:
maybe removing the locks and then replacing the line 

```C#
pSortHandle->collatorsPerOption[options] = pCollator;
```

with 

```C#
            if (!__sync_bool_compare_and_swap(&pSortHandle->collatorsPerOption[options], NULL, pCollator))
            {
                ucol_close(pCollator);
                pCollator = pSortHandle->collatorsPerOption[options];
                assert(pCollator != NULL && "pCollator not expected to be null here.");
            }

```
would be enough. 

@jkotas can advise if we can use __sync_bool_compare_and_swap though here.

Reply:
`__sync_bool_compare_and_swap` is GCC/LLVM-specific intrinsic.

If possible, we should use standard C methods from stdatomic.h: https://en.cppreference.com/w/c/atomic/atomic_compare_exchange

Reply:
I have tried to `use atomic_compare_exchange_strong` from `<stdatomic.h>` but the compiler was giving me following error:

```log
error: address argument to atomic operation must be a pointer to _Atomic type ('UCollator *' (aka 'struct UCollator *') invalid)
```

So I have used `__sync_bool_compare_and_swap` as suggested by @tarekgh 

It compiles fine and produces following ASM on my Ubuntu 16.04 x64 vm:

```asm
    3d72:	48 89 c1             	mov    %rax,%rcx
    3d75:	31 c0                	xor    %eax,%eax
    3d77:	f0 48 0f b1 0b       	lock cmpxchg %rcx,(%rbx)
    3d7c:	74 14                	je     3d92 <GetCollatorFromSortHandle+0x42>
```

Reply:
> address argument to atomic operation must be a pointer to _Atomic type

You need to wrap the pointer inside `_Atomic` to fix this, like `_Atomic(UCollator*)`.

> It compiles fine and produces following ASM on my Ubuntu 16.04 x64 vm

And does it produce the correct code on other platforms too? These low-level GCC/LLVM intrinsics typically need to be augmented with explicit barriers for other platforms to do what you need.